File: fletcher8.pas
unit fletcher8;
//Fletcher-8
//Author: domasz
//Version: 0.1 (2022-11-17)
//Licence: MIT

interface

uses SysUtils, HasherBase;

type THasherFletcher8 = class(THasherbase)
  private
    FHash, FHash2: Byte;
  public
    constructor Create; override;
    procedure Update(Msg: PByte; Length: Integer); override;
    function Final: String; override;
end;

implementation

constructor THasherFletcher8.Create;
begin
  inherited Create;
  Check := '0C';
  FHash := 0;
  FHash2 := 0;
end;

procedure THasherFletcher8.Update(Msg: PByte; Length: Integer);
var i: Integer;
begin
  for i:=0 to Length-1 do begin
    FHash := (FHash + Msg^) mod 15;
    FHash2 := (FHash2 + FHash) mod 15;

    Inc(Msg);
  end;   
end;

function THasherFletcher8.Final: String;
begin
  FHash := (FHash2 shl 4) or FHash;

  Result := IntToHex(FHash, 2);
end;

initialization
  HasherList.RegisterHasher('Fletcher-8', THasherFletcher8);

end.

File: PHP/Algos/fletcher8.php
<?php
//Fletcher-32
//Author: domasz
//Version: 0.1 (2022-11-19)
//Licence: MIT

require_once 'HasherBase.php';

class HasherFletcher8 extends HasherBase
{
	protected $FHash, $FHash2 = 0;

	public function __construct() 
	{
		$this->Check = '0C';
		$this->FHash = 0;
		$this->FHash2 = 0;
	}

	public function Update($Msg, $Length)
	{
		for ($i=0; $i<$Length; $i++)
		{
			$this->FHash = ($this->FHash + ord($Msg[$i]) ) % 15;
			$this->FHash2 = ($this->FHash2 + $this->FHash) % 15;
		}
	}

	public function Finish()
	{
		$this->FHash = ($this->FHash2 << 4) | $this->FHash;
		return sprintf('%02X', $this->FHash);
	}
}

$HasherList->RegisterHasher('Fletcher-8', 'HasherFletcher8');